src/mono/mono/metadata/assembly.c

```diff
@@ -1470,44 +1471,46 @@ bundled_assembly_match (const char *bundled_name, const char *name)
 }
 
 static MonoImage *
-open_from_bundle_internal (MonoAssemblyLoadContext *alc, const char *filename, MonoImageOpenStatus *status, gboolean is_satellite)
+open_from_bundle_internal (MonoAssemblyLoadContext *alc, const char *filename, MonoImageOpenStatus *status)
 {
-	if (!bundles)
+	if (!bundles && !mono_bundled_resource_get_bundle_contains_assemblies ())
 		return NULL;
 
-	MonoImage *image = NULL;
-	char *name = is_satellite ? g_strdup (filename) : g_path_get_basename (filename);
-	for (int i = 0; !image && bundles [i]; ++i) {
-		if (bundled_assembly_match (bundles[i]->name, name)) {
+	MonoBundledAssemblyResource *assembly = (MonoBundledAssemblyResource *)mono_bundled_resource_get_bundled_resource (filename);
+	if (assembly)
+		return mono_image_open_from_data_internal (alc, (char *)assembly->assembly.data, assembly->assembly.size, FALSE, status, FALSE, filename, NULL);
+
+	for (int i = 0; bundles [i]; ++i) {
+		if (bundled_assembly_match (bundles[i]->name, filename)) {
 			// Since bundled images don't exist on disk, don't give them a legit filename
-			image = mono_image_open_from_data_internal (alc, (char*)bundles [i]->data, bundles [i]->size, FALSE, status, FALSE, name, NULL);
-			break;
+			return mono_image_open_from_data_internal (alc, (char*)bundles [i]->data, bundles [i]->size, FALSE, status, FALSE, filename, NULL);
 		}
 	}
 
-	g_free (name);
-	return image;
+	return NULL;
 }
 
 static MonoImage *
 open_from_satellite_bundle (MonoAssemblyLoadContext *alc, const char *filename, MonoImageOpenStatus *status, const char *culture)
 {
-	if (!satellite_bundles)
+	if (!satellite_bundles && !mono_bundled_resource_get_bundle_contains_satellite_assemblies ())
 		return NULL;
 
 	MonoImage *image = NULL;
-	char *name = g_strdup (filename);
+	char *bundle_name = g_strconcat (culture, "/", filename, (const char *)NULL);
+
+	MonoBundledSatelliteAssemblyResource *satellite_assembly = (MonoBundledSatelliteAssemblyResource *)mono_bundled_resource_get_bundled_resource (filename);
```

Comment:
Or is `filename` in the case that this is coming from the resource bundle already a path, not just a file name? Is the registeredName for a satellite assembly something like "es-ES/System.Runtime.resources.dll" ?

In that case, I don't understand how this lookup could ever succeed --- I'm pretty sure `open_from_satellite_bundle` is passed the filename (not a relative path) and culture separately.

did I get that wrong?

Is there a way to test this out?

Reply:
Looking at the old way of handling it, `mono_assembly_open_from_bundle` will call `open_from_satellite_bundle` if it gets an open call that includes a culture string. It will then search all registered bundled satellite assemblies based on name and culture, if it finds a match it will then concat, `culture + / + name` and use that as the name in call to `mono_image_open_from_data_internal`, representing a virtual path to the resource on disk.

How we index the satellite assembly in our hash table is something we can decide internal, since we still end up concat `culture + / + name` I think we could just use that as the hash key of the resource in the hash map, making it unique. The lookup would then search using the `culture + / + name` right away and use that in call to `mono_image_open_from_data_internal`. In order to remove need to duplicate strings when register the bundle to build the hash key, maybe we should extend MonoBundledResource with a const char *path field, for most resources we currently have, it will be the same as name, but in cases where it represent something else, like a relative file path as for satellite assemblies, it will be the unique name `culture + / + name`. Having that in a separate field, makes it possible to pre-allocate it and reduce need to do string allocation when register a resource.

Reply:
I agree that we should have a `path` and a `name`.

I'm not totally sure about completely dropping culture and just basing everything on a virtual path naming convention.

Reply:
Then maybe we should add a virtual path field to all resources, and keep name and culture as part of satellite assembly type. Lookups will then be done against the path and not name, but for most cases, resource will have same path as name.

Reply:
I wasn't sure exactly how satellite assemblies would look like through the library builder, mono aot compiler, and respective app builders, so I didn't foresee that the satellite assemblies would have the same name as each other.
I naively thought that they would be `<satellite1>.resources.dll` `<satellite2>.resources.dll` etc instead of being `<culture1>/<satellite>.resources.dll` `<culture2>/<satellite>.resources.dll` ...

I added an extra field to the base `MonoBundledResources` struct to hold the unique ID of the resource that would be the key added to the hash table.
For all resources that are not satellite assemblies, it would be the same as the `name` field that is used in the respective `MonoBundledAssembly` or `MonobundledData` struct, but for satellite_assemblies, it will be `<culture>/<name>` as suggested above.
Moreover, the `name` for the satellite assemblies (`MonoBundledSatelliteAssembly->name`) is still just the `name`, and not a modification to include the culture, and `MonoBundledSatelliteAssembly->culture` is retained as well.
The lookup to search through bundled satellite assemblies will be based on `<culture>/<name>` (i.e. bundle_name) instead of the `<name>` now.